### PR TITLE
ACR Dashboard Subtle Notification

### PR DESCRIPTION
**Pull request by benbowler**
## Summary

Addresses issue:

- #9371 

## Relevant technical choices

PR grew a bit bigger, since this is the main banner issue that connects previous implementations. I covered some edge cases, and made some tweaks on existing implementation 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

**Review by 10upsimon — CHANGES_REQUESTED** [review]
@zutigrm while the logic works well and the banner is triggered when I test via local state manipulation, I do see the following console error after actioning the banner (i.e Add Metrics):

```
react-dom.development.js:88 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in ConversionReportingNotificationCTAWidget (created by WhenAnalytics4Active(ConversionReportingNotificationCTAWidget))
    in WhenAnalytics4Active(ConversionReportingNotificationCTAWidget) (created by WidgetRenderer)
    in div (created by ForwardRef)
    in div (created by ForwardRef)
    in ForwardRef (created by WithWidgetSlug)
    in WithWidgetSlug (created by WidgetRenderer)
    in WidgetRenderer (created by WidgetAreaRenderer)
    in WidgetErrorHandler (created by WidgetAreaRenderer)
    in div (created by Cell)
    in Cell (created by WidgetCellWrapper)
```

Let's get that fixed and then I think this is good to go.

**Review by 10upsimon — APPROVED** [review]
@zutigrm latest changes have solved the console error for me, nice work ✅

**Comment by eugene-manuilov** [comment]
@zutigrm this PR contains a lot of changes that are not related to 9371. Why did you add all of these extra changes here? Can you create a new PR that just adds what is needed to implement the new notification as mentioned in 9371? Everything else can be reviewed separately.

File: assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		isUserInputCompleted &&
+		haveConversionReportingEventsForTailoredMetrics &&
+		! keyMetricSettings?.includeConversionTailoredMetrics;
+
+	const { setKeyMetricsSetting, saveKeyMetricsSettings } =
+		useDispatch( CORE_USER );
+	const { dismissNewConversionReportingEvents } =
+		useDispatch( MODULES_ANALYTICS_4 );
+
+	const handleCTAClick = useCallback( () => {
```
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js:75]: Let's rename this to be `handleAddMetricsClick`.

File: assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	}, [
+		setKeyMetricsSetting,
+		saveKeyMetricsSettings,
+		dismissNewConversionReportingEvents,
+		shouldShowInitialCalloutForTailoredMetrics,
+	] );
+
+	const onDismiss = useCallback( () => {
+		dismissNewConversionReportingEvents();
+	}, [ dismissNewConversionReportingEvents ] );
```
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js:95]: This callback seems redundant, no? Let's remove it.

File: assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import { Button, SpinnerButton } from 'googlesitekit-components';
+import StarFill from '../../../svg/icons/star-fill.svg';
+import SubtleNotification from '../../googlesitekit/notifications/components/layout/SubtleNotification';
+import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
+import Link from '../Link';
+
+export default function ConversionReportingDashboardSubtleNotification( {
+	ctaLabel,
+	handleCTAClick,
+	isSaving,
```
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js:43]: `isSaving` is not required, we need to set the default value for it:
```suggestion
	isSaving = false,
```

File: assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * Internal dependencies
+ */
+import { useSelect } from 'googlesitekit-data';
+import { Button, SpinnerButton } from 'googlesitekit-components';
+import StarFill from '../../../svg/icons/star-fill.svg';
+import SubtleNotification from '../../googlesitekit/notifications/components/layout/SubtleNotification';
+import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
+import Link from '../Link';
+
+export default function ConversionReportingDashboardSubtleNotification( {
```
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js:40]: Maybe let's rename this component to be a bit more concise? 
```suggestion
export default function ConversionEventsNotification( {
```

File: assets/js/googlesitekit/widgets/register-defaults.js [hunk 5]
```diff
@@ -467,15 +468,23 @@ export function registerDefaults( widgetsAPI ) {
...
-				isActive: () => {
+				isActive: ( select ) => {
 					if ( ! isFeatureEnabled( 'conversionReporting' ) ) {
 						return false;
 					}
 
+					if (
+						! select(
+							MODULES_ANALYTICS_4
+						).hasNewConversionReportingEvents()
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/widgets/register-defaults.js:483]: Before calling this selector we need to ensure that the Analytics module is connected.

File: assets/js/googlesitekit/datastore/user/key-metrics.js [hunk 6]
```diff
@@ -451,28 +501,40 @@ const baseSelectors = {
 	 *
 	 * @return {Array<string>|undefined} An array of Key Metric widget slugs, or undefined if the user input settings are not loaded.
 	 */
+	/* eslint-disable complexity */
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/googlesitekit/datastore/user/key-metrics.js:504]: Yes, forgot I left it there before it was split in another PR

File: assets/js/googlesitekit/datastore/user/constants.js [hunk 7]
```diff
@@ -124,3 +124,34 @@ export const allKeyMetricsTileWidgets = [
...
+
+/**
+ * Gets conversion events related metrics.
+ *
+ * @since n.e.x.t
+ * @private
+ *
+ * @return {Object} Metrics list object.
+ */
+export function getKeyMetricsConversionEventWidgets() {
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/googlesitekit/datastore/user/constants.js:136]: Moved it to the new util function, since hook is moved to a selector

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 8]
```diff
@@ -157,11 +158,33 @@ export default function UserInputQuestionnaire() {
...
 	const submitChanges = useCallback( async () => {
 		trackEvent( gaEventCategory, 'summary_submit' );
 
 		const response = await saveUserInputSettings();
 		if ( ! response.error ) {
+			// If selected purpose has any ACR KMW assigned to it,
+			// mark 'includeConversionTailoredMetrics' key metrics setting to true
+			// so they can be included.
+			if ( haveConversionReportingEventsForTailoredMetrics ) {
+				await setKeyMetricsSetting(
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/components/user-input/UserInputQuestionnaire.js:179]: It is not necessary actually, removed

File: assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	if ( ! shouldShowInitialCalloutForTailoredMetrics ) {
+		return <WidgetNull />;
+	}
+
+	return (
+		<Widget noPadding fullWidth>
+			<ConversionReportingDashboardSubtleNotification
+				ctaLabel={ __( 'Add metrics', 'google-site-kit' ) }
+				handleCTAClick={ () => handleCTAClick() }
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js:105]: Oh yes, I was moving some argument, temporarily and then forgot to switch back. Thanks, edited

File: assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js (hunk 2 above)
Comment (zutigrm) [reply assets/js/components/KeyMetrics/ConversionReportingNotificationCTAWidget.js:95]: Agree, removed

File: assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js (hunk 4 above)
Comment (zutigrm) [reply assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js:40]: It is more on matching the naming convention for other callout used in the settings https://github.com/google/site-kit-wp/blob/440dcd6e8289a8776acc60ecb3c13f99bfd0ac5a/assets/js/components/KeyMetrics/ConversionReportingSettingsSubtleNotification.js#L34

File: assets/js/googlesitekit/widgets/register-defaults.js (hunk 5 above)
Comment (zutigrm) [reply assets/js/googlesitekit/widgets/register-defaults.js:483]: Is that necessary? This widget is hooked into key metrics widget area, the widget context of that area will not be active until KM widget area is shown - which is shown once GA4 module is connected, so we should be pretty safe this selector won;t be invoked otherwise

File: assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js (hunk 4 above)
Comment (eugene-manuilov) [reply assets/js/components/KeyMetrics/ConversionReportingDashboardSubtleNotification.js:40]: Ok, let it be it.

**Comment by zutigrm** [comment]
Thanks @eugene-manuilov PR updated

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, a few more comments from me.

File: assets/js/modules/analytics-4/datastore/conversion-reporting.js [hunk 10]
```diff
@@ -226,6 +254,49 @@ export const selectors = {
...
+	 * Checks if there are key metrics widgets connected with the detected events for the supplied purpose answer.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}  purpose      Value of saved site purpose from user input settings.
+	 * @param {boolean} useNewEvents Flag inclusion of detected new events, otherwise initial detected events will be used.
+	 * @return {boolean} TRUE if current site purpose will have any ACR key metrics widgets assigned to it, FALSE otherwise.
+	 */
+	haveConversionEventsForTailoredMetrics: createRegistrySelector(
+		( select ) => ( state, purpose, useNewEvents ) => {
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/modules/analytics-4/datastore/conversion-reporting.js:268]: I removed `purpose` since currently it won't have usage, but we need `useNewEvents` as following issue that depends on this one will use that, and core infrastructure is laid down here

**Comment by zutigrm** [comment]
@eugene-manuilov Thanks, latest comments are applied to the PR

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 11]
```diff
@@ -157,11 +158,33 @@ export default function UserInputQuestionnaire() {
 		questionNumber,
 	] );
 
+	const haveConversionReportingEventsForTailoredMetrics = useSelect(
+		( select ) =>
+			select(
+				MODULES_ANALYTICS_4
+			).haveConversionEventsForTailoredMetrics()
+	);
```
Comment (eugene-manuilov) [inline assets/js/components/user-input/UserInputQuestionnaire.js:166]: I think here we also need to ensure that the Analytics module is connected.

File: assets/js/components/KeyMetrics/ConfirmSitePurposeChangeModal.js [hunk 12]
```diff
@@ -48,15 +48,23 @@ import {
...
 
 function ConfirmSitePurposeChangeModal( {
 	dialogActive = false,
 	handleDialog = null,
 } ) {
 	const [ isSaving, setIsSaving ] = useState( false );
 
+	const includeConversionTailoredMetrics = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).haveConversionEventsForTailoredMetrics()
+	);
```
Comment (eugene-manuilov) [inline assets/js/components/KeyMetrics/ConfirmSitePurposeChangeModal.js:61]: And here as well, we need to ensure that the Analytics module is connected before calling its selector.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. A few more comments from me and it should be good to go.

File: assets/js/modules/analytics-4/datastore/conversion-reporting.js [hunk 13]
```diff
@@ -226,6 +263,81 @@ export const selectors = {
...
+
+	/**
+	 * Gets conversion events related metrics.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @return {Object} Metrics list object.
+	 */
+	getKeyMetricsConversionEventWidgets() {
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply assets/js/modules/analytics-4/datastore/conversion-reporting.js:319]: You are correct, it won't be used in other places if I recall correctly, will move it within the selector

**Review by eugene-manuilov — APPROVED** [review]
Thanks, LGTM.
